### ISSUE TITLE
duplicate subtitles

### ISSUE THREAD

**Issue by nzbjamz** [mdhiggins/sickbeard_mp4_automator#655]
I'm seeing subtitles get added to a video that already has an english (MOV_TEXT) subtitle. The result is when it's trancoded there are two (MOV_TEXT) english subtitles. It also looks like there's duplicate audio tracks (3 aac stereo; 1 ac3 5.1).

My settings are:
```txt
[MP4]
audio-default-language = eng
subtitle-language = eng
subtitle-default-language = eng
convert-mp4 = True
tag-language = eng
download-subs = True
```
Here is the output of the `manual.py -a -i ./` run:
```shell
Audio bitrate was invalid, defaulting to 256 per channel.
64bit Python.
Processing file x.mp4
Matched x
Processing x
Reading video stream.
Video codec detected: h264.
Pix Fmt: yuv420p.
Reading audio streams.
Audio detected for stream #1: aac [eng].
Creating audio stream 0 from source stream 1.
Audio Track is number 1 setting disposition to default
Audio detected for stream #2: aac [eng].
Creating audio stream 1 from source stream 2.
Audio Track is number 2 setting disposition to none
Audio detected for stream #3: ac3 [eng].
Creating audio stream 2 from source audio stream 3 [iOS-audio].
Audio track is number 2 setting disposition to none
Creating audio stream 3 from source stream 3.
Audio Track is number 3 setting disposition to none
Reading subtitle streams.
Subtitle detected for stream #4: mov_text [eng].
Creating subtitle stream 0 from source stream 4.
Attempting to download subtitles.
External eng subtitle file detected.
Creating subtitle stream 1 by importing x.en.srt.
{
    "preopts": [
        "-fix_sub_duration"
    ],
    "subtitle": {
        "0": {
            "map": 4,
            "codec": "mov_text",
            "language": "eng",
            "encoding": null
        },
        "1": {
            "path": "/share/CACHEDEV1_DATA/Multimedia/x.en.srt",
            "map": 0,
            "codec": "mov_text",
            "language": "eng",
            "source": 1
        }
    },
    "format": "mp4",
    "postopts": [
        "-threads",
        "auto"
    ],
    "video": {
        "map": 0,
        "codec": "copy",
        "bitrate": 4587.6886999999997,
        "level": 4.0
    },
    "audio": {
        "0": {
            "map": 1,
            "language": "eng",
            "channels": 2,
            "codec": "copy",
            "disposition": "default",
            "bitrate": 256,
            "filter": null
        },
        "1": {
            "map": 2,
            "language": "eng",
            "channels": 2,
            "codec": "copy",
            "disposition": "none",
            "bitrate": 256,
            "filter": null
        },
        "2": {
            "map": 3,
            "language": "eng",
            "channels": 2,
            "codec": "aac",
            "disposition": "none",
            "bitrate": 256,
            "filter": null
        },
        "3": {
            "map": 3,
            "language": "eng",
            "channels": 6,
            "codec": "copy",
            "disposition": "none",
            "bitrate": 1536,
            "filter": null
        }
    }
}
Starting conversion...
```

**Comment by nzbjamz** [comment]
Running `manual.py -a -i x.mp4` on the same `x.mp4` multiple times just keeps adding audio tracks too.
Here's another run:
```shell
    "audio": {
        "0": {
            "map": 1,
            "language": "eng",
            "channels": 2,
            "codec": "copy",
            "disposition": "default",
            "bitrate": 256,
            "filter": null
        },
        "1": {
            "map": 2,
            "language": "eng",
            "channels": 2,
            "codec": "copy",
            "disposition": "none",
            "bitrate": 256,
            "filter": null
        },
        "2": {
            "map": 3,
            "language": "eng",
            "channels": 2,
            "codec": "copy",
            "disposition": "none",
            "bitrate": 256,
            "filter": null
        },
        "3": {
            "map": 4,
            "language": "eng",
            "channels": 2,
            "codec": "aac",
            "disposition": "none",
            "bitrate": 256,
            "filter": null
        },
        "4": {
            "map": 4,
            "language": "eng",
            "channels": 6,
            "codec": "copy",
            "disposition": "none",
            "bitrate": 1536,
            "filter": null
        }
    }
```
and if I run it again:
```shell
"audio": {
        "0": {
            "map": 1,
            "language": "eng",
            "channels": 2,
            "codec": "copy",
            "disposition": "default",
            "bitrate": 256,
            "filter": null
        },
        "1": {
            "map": 2,
            "language": "eng",
            "channels": 2,
            "codec": "copy",
            "disposition": "none",
            "bitrate": 256,
            "filter": null
        },
        "2": {
            "map": 3,
            "language": "eng",
            "channels": 2,
            "codec": "copy",
            "disposition": "none",
            "bitrate": 256,
            "filter": null
        },
        "3": {
            "map": 4,
            "language": "eng",
            "channels": 2,
            "codec": "copy",
            "disposition": "none",
            "bitrate": 256,
            "filter": null
        },
        "4": {
            "map": 5,
            "language": "eng",
            "channels": 2,
            "codec": "aac",
            "disposition": "none",
            "bitrate": 256,
            "filter": null
        },
        "5": {
            "map": 5,
            "language": "eng",
            "channels": 6,
            "codec": "copy",
            "disposition": "none",
            "bitrate": 1536,
            "filter": null
        }
    }
```

**Comment by nzbjamz** [comment]
I'm guessing there's a bug in the ios-audio bit and it's always adding an extra channel (even when it doesn't need to).

**Comment by nzbjamz** [comment]
If there's any way to clean these up, that would be really nice too.

**Comment by Rycore** [comment]
Hey @nzbjamz, I'm new to this but it looks like instead of ignoring the download-subs option when subtitles already exist, it's downloading any existing subtitle(s) regardless.

```
Subtitle detected for stream #4: mov_text [eng].
Creating subtitle stream 0 from source stream 4.
Attempting to download subtitles.
```

The only way to sort this would for the dev to include an option to disable or ignore download-subs, or if you wrote a script to use ffprobe to detect if subtitles (in your chosen language) exists and then modify autoProcess.ini to enable or disable download-subs.

I'm also curious to know, is 
```
[MP4]
audio-default-language = eng
subtitle-language = eng
subtitle-default-language = eng
convert-mp4 = True
tag-language = eng
download-subs = True
```
all of your settings in autoProcess.ini? because if you've removed all the settings you don't need, this could be causing the problem with the duplicate audio tracks.

**Comment by Rycore** [comment]
Also, have you tried running manual.py without auto mode? just to see if the problem still occurs.

**Comment by mdhiggins** [comment]
Its stated in the readme and on multiple issues if you're going to reprocess files created by the mp4 automator you need to make sure iOS audio is disabled or it will continued to recreate the iOS audio steam. As of now there is no way for the script to reliably determine that one track is simply a 2 channel version of another track, so it has to assume you still wanted to create a stereo track since the option is enabled.

As far as the subtitles go, I can't specifically comment there. The download-subs feature uses subliminal. Subliminal is supposed to analyze and determine if appropriate subtitle are already present and then avoid redownloading them.

You can disable download-subs if you're finding this option to be inconsistent, but if you want to improve that reliability you'll need to speak to the developers of subliminal.

I am curious why you're reprocessing the same mp4 files multiple times? What end goal are you attempting to achieve? If you're just attempting to shoehorn automation by calling the manual.py script on your entire media library every time you add a new file, I would advise you either don't do this or you disable the convert-mp4 option. Disabling the convert mp4 option would fix all these issues as it won't attempt to download subs or convert any streams and will simply update the metadata with the latest and move on.

**Comment by nzbjamz** [comment]
@Rycore
> I'm also curious to know, is [this] all of your settings in autoProcess.ini? 

No. It's just the bits I think are relevant.

> it looks like instead of ignoring the download-subs option when subtitles already exist, it's downloading any existing subtitle(s) regardless.

Ah ya it could be. The comment in the code suggests
```python
# Attempt to download subtitles if they are missing using subliminal
```
I was wanting that behavior not a forced download each time.

> Also, have you tried running manual.py without auto mode? just to see if the problem still occurs.

I haven't. I want to apply the conversion settings across several files, so performing the manual conversion 1 by 1 is not something to do here.

**Comment by nzbjamz** [comment]
@mdhiggins 
> The download-subs feature uses subliminal. Subliminal is supposed to analyze and determine if appropriate subtitle are already present and then avoid redownloading them.

Maybe the extra iphone-audio channel is throwing it off?

> If you're just attempting to shoehorn automation by calling the manual.py script on your entire media library every time you add a new file, I would advise you either don't do this or you disable the convert-mp4 option.

Not shoehorning automation. I'm using it to fix things like `h264-max-level` or add missing subs.

**Comment by mdhiggins** [comment]
If you're reprocessing your library you just need to temporarily turn off the iOS-audio option and that will fix your endlessly replicated audio stream issue. No other good way to trim the streams without causing problems in other more common scenarios. You can make an alternative config file with that off and point manual.py to that config when reprocessing existing files for this purpose.

The attempting to download subtitles message just means subliminal is analyzing the files, I could probably change that to reflect it. On my todo list is to update to subliminal version 2 which may have some improvements to this behavior but not downloading additional subs is something this version of subliminal is supposed to do.

**Comment by Rycore** [comment]
Okay so I tested this and I still having the same issue, I decided to run this file through my script to see if it produced the same issue and it did, so after looking into it a bit further, it turns out that there seems to be an issue with FFmpeg were if a file contains chapters, it will add it as a subtitle.

Now someone else has reported this issue but with this other person, it seems to be happening when he's copying subtitles over, whereas I've not had that issue (to my knowledge) if subtitles already exist within the file.

Here is the link to his post [ffmpeg.gusari.org]( http://ffmpeg.gusari.org/viewtopic.php?f=11&t=2191)

So I was mistaken, it appears to be FFmpeg which is causing the problem, I'm going to look into this more and see if there is a workaround as I currently have the latest (as the mod in that forum suggest the OP download) and to re-build FFmpeg with an older version would be a pain considering it takes about 2 hours to do, anyway if I find a workaround, possibly adding **-sn** to the script to disable subtitles if none exist or something like that, I'll post back here within the next couple of hours.

I should also note that I tested two other files, an .avi and .mkv with my script to make sure there were no bugs, so it is definitely due to the chapters in the video which is causing this issue.
